src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs

```diff
@@ -0,0 +1,489 @@
+    /// </para>
+    /// </remarks>
+    public class JsonEventFormatter : CloudEventFormatter
+    {
+        private const string JsonMediaType = "application/json";
+        private const string MediaTypeSuffix = "+json";
+
+        /// <summary>
+        /// The property name to use for base64-encoded binary data in a structured-mode message.
+        /// </summary>
+        protected const string DataBase64PropertyName = "data_base64";
+
+        /// <summary>
+        /// The property name to use for general data in a structured-mode message.
+        /// </summary>
+        protected const string DataPropertyName = "data";
+
+        /// <summary>
+        /// The options to use when serializing objects to JSON.
+        /// </summary>
+        protected JsonSerializerOptions SerializerOptions { get; }
+
+        /// <summary>
+        /// The options to use when parsing JSON documents.
+        /// </summary>
+        protected JsonDocumentOptions DocumentOptions { get; }
+
+        /// <summary>
+        /// Creates a JsonEventFormatter that uses a default <see cref="JsonSerializer"/>.
+        /// </summary>
+        public JsonEventFormatter() : this(null, default)
+        {
+        }
+
+        /// <summary>
+        /// Creates a JsonEventFormatter that uses the specified <see cref="JsonSerializer"/>
+        /// to serialize objects as JSON.
+        /// </summary>
+        /// <param name="serializerOptions">The options to use when serializing objects to JSON. May be null.</param>
+        /// <param name="documentOptions">The options to use when parsing JSON documents.</param>
+        public JsonEventFormatter(JsonSerializerOptions serializerOptions, JsonDocumentOptions documentOptions)
+        {
+            SerializerOptions = serializerOptions;
+            DocumentOptions = documentOptions;
+        }
+
+        public override async Task<CloudEvent> DecodeStructuredModeMessageAsync(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
+        {
+            data = data ?? throw new ArgumentNullException(nameof(data));
+            var encoding = contentType.GetEncoding();
+
+            JsonDocument document;
+            if (encoding is UTF8Encoding)
+            {
+                document = await JsonDocument.ParseAsync(data, DocumentOptions).ConfigureAwait(false);
+            }
+            else
+            {
+                using var reader = new StreamReader(data, encoding);
+                string json = await reader.ReadToEndAsync().ConfigureAwait(false);
+                document = JsonDocument.Parse(json, DocumentOptions);
+            }
+            using (document)
+            {
+                return DecodeJsonDocument(document, extensionAttributes);
+            }
+        }
+
+        public override CloudEvent DecodeStructuredModeMessage(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
```

Comment:
nit - consider calling into the async method and passing a `bool async` param that can be used to check whether to call sync or async methods.

Reply:
I've done this in a separate commit - see whether it's what you were expecting. It looks nice other than the exception unwrapping part... if you're aware of a framework method that does a sort of "give me the task result, unwrapping any exceptions appropriately" that would make it more of a slam dunk change.

Reply:
Doh - `GetAwaiter().GetResult()` is fine. PTAL - if this is what you expected, I'll rebase it into the previous commit, and merge.